airflow/timetables/interval.py

```diff
@@ -43,6 +43,13 @@ def __eq__(self, other: Any) -> bool:
     def validate(self) -> None:
         self._schedule.validate()
 
+    def infer_data_interval(self, run_after: DateTime) -> Optional[DataInterval]:
+        # Get the last complete period before run_after, e.g. if a DAG run is
+        # scheduled at each midnight, the data interval of a manually trioggered
+        # run at 1am 25th is between 0am 24th and 0am 25th.
```

Comment:
(Without thinking about this in much detail) I would have expected such a manually triggered run to be from midnight 25th to 1am 25th.

But maybe this way does make more sense. Data intervals aren't (currently) well defined in the case of manual triggers.

Reply:
What I like about this strategy is we set data interval to a period that always makes sense, instead of some in-between period that doesn’t conform to any DAG scheduling rules. But admittedly I don’t really know what people use manual triggers for (except for testing, for which the data interval doesn’t matter anyway) so maybe this is not what they want.

Reply:
Options:

(1) Shouldn't the `data_interval_start_date` be 24th 1 am and `data_interval_end_date` be 25th 1am so `data_interval_end_date` = `schedule_data` / `run_date`. Why the offset of 1 hour?

Reply:
If the DAG is expected to be scheduled for `0 0 * * *`, data intervals always start and end at midnight for all scheduled (and backfilled) runs, so a manual run should match that. If the interval is `timedelta(days=1)` then yes this manual run’s data interval should be between 1am 24th and 1am 25th. I will modify the docstring to clarify this.